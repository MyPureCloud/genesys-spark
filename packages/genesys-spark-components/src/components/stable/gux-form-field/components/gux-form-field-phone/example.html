<h1>gux-form-field-phone</h1>
<form onchange="notify(event)" oninput="notify(event)">
  <fieldset>
    <legend>No Default Region</legend>
    <div class="field-container">
      <gux-form-field-phone label-position="above">
        <gux-phone-input-beta></gux-phone-input-beta>
        <label slot="label">Phone Input</label>
      </gux-form-field-phone>
    </div>
    <div class="field-container">
      <gux-form-field-phone label-position="above">
        <gux-phone-input-beta required="true"></gux-phone-input-beta>
        <label slot="label">Phone Input - Required</label>
      </gux-form-field-phone>
    </div>
    <div class="field-container">
      <gux-form-field-phone label-position="above">
        <gux-phone-input-beta disabled="true"></gux-phone-input-beta>
        <label slot="label">Phone Input - Disabled</label>
      </gux-form-field-phone>
    </div>
    <div class="field-container">
      <gux-form-field-phone label-position="above">
        <gux-phone-input-beta has-error></gux-phone-input-beta>
        <label slot="label">Phone Input - Custom Error</label>
        <span slot="error">Custom Error Message</span>
      </gux-form-field-phone>
    </div>
    <div class="field-container">
      <gux-form-field-phone label-position="above">
        <gux-phone-input-beta></gux-phone-input-beta>
        <label slot="label">Phone Input - Custom Hint</label>
        <span slot="help">Custom Help Message</span>
      </gux-form-field-phone>
    </div>
  </fieldset>
  <fieldset>
    <legend>Default Region Set</legend>
    <div class="field-container">
      <gux-form-field-phone label-position="above">
        <gux-phone-input-beta default-region="JP"></gux-phone-input-beta>
        <label slot="label">Phone Input - Default Region = Japan</label>
      </gux-form-field-phone>
    </div>
    <div class="field-container">
      <gux-form-field-phone label-position="above">
        <gux-phone-input-beta default-region="DE"></gux-phone-input-beta>
        <label slot="label">Phone Input - Default Region = Germany</label>
      </gux-form-field-phone>
    </div>
    <div class="field-container">
      <gux-form-field-phone label-position="above">
        <gux-phone-input-beta default-region="CH"></gux-phone-input-beta>
        <label slot="label">Phone Input - Default Region = Switzerland</label>
      </gux-form-field-phone>
    </div>
  </fieldset>
</form>

<style>
  .field-container {
    margin-bottom: 12px;
  }

  fieldset {
    margin-bottom: 24px;
  }
</style>
