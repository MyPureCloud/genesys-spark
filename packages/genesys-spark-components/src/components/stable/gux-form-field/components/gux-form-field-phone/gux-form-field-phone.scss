@use '../../gux-form-field.scss';
@use '../../functional-components/gux-form-field-fieldset-container/gux-form-field-fieldset-container.scss';
@use '../../functional-components/gux-form-field-error/gux-form-field-error.scss';
@use '../../functional-components/gux-form-field-legend-label/gux-form-field-legend-label.scss';

@include gux-form-field-fieldset-container.Style;
@include gux-form-field-error.Style;
@include gux-form-field-legend-label.Style;

::slotted(label) {
  @include gux-form-field.gse-ui-formControl-label;
}

.gux-input-and-error-container {
  flex-grow: 1;
}
