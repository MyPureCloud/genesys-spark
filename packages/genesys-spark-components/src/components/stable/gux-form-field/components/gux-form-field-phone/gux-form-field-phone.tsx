import {
  Component,
  Element,
  h,
  JSX,
  Listen,
  Prop,
  State,
  Watch
} from '@stencil/core';

import { buildI18nForComponent, GetI18nValue } from '../../../../../i18n';
import { ILocalizedComponentResources } from '../../../../../i18n/fetchResources';
import { OnMutation } from '@utils/decorator/on-mutation';
import {
  onDisabledChange,
  onRequiredChange
} from '@utils/dom/on-attribute-change';
import { trackComponent } from '@utils/tracking/usage';

import {
  GuxFormFieldError,
  GuxFormFieldFieldsetContainer,
  GuxFormFieldScreenreaderLabel,
  GuxFormFieldVisualLabel
} from '../../functional-components/functional-components';
import {
  GuxFormFieldLabelPosition,
  GuxFormFieldIndicatorMark
} from '../../gux-form-field.types';
import { hasSlot } from '@utils/dom/has-slot';
import { getSlotTextContent } from '@utils/dom/get-slot-text-content';
import {
  getComputedLabelPosition,
  validateFormIds
} from '../../gux-form-field.service';

import componentResources from './i18n/en.json';
import { GuxFormFieldHelp } from '../../functional-components/gux-form-field-help/gux-form-field-help';

/**
 * @slot - Required slot for gux-phone-input-beta tag
 * @slot label - Required slot for label tag
 * @slot error - Optional slot for error message
 * @slot help - Optional slot for help message
 * @slot label-info - Optional slot for label tooltip
 */
@Component({
  styleUrl: 'gux-form-field-phone.scss',
  tag: 'gux-form-field-phone',
  shadow: true
})
export class GuxFormFieldPhone {
  private getI18nValue: GetI18nValue;
  private input: HTMLGuxPhoneInputBetaElement;
  private label: HTMLLabelElement;
  private labelInfo: HTMLGuxLabelInfoBetaElement;
  private disabledObserver: MutationObserver;
  private requiredObserver: MutationObserver;
  private hideLabelInfoTimeout: ReturnType<typeof setTimeout>;

  @Element()
  private root: HTMLElement;

  @Prop()
  labelPosition: GuxFormFieldLabelPosition;

  /**
   * Field indicator mark which can show *, (optional) or blank
   * Defaults to required. When set to required, the component will display * for required fields and blank for optional
   * When set to optional, the component will display (optional) for optional and blank for required.
   */
  @Prop()
  indicatorMark: GuxFormFieldIndicatorMark = 'required';

  @State()
  private computedLabelPosition: GuxFormFieldLabelPosition = 'above';

  @State()
  private disabled: boolean;

  @State()
  private required: boolean = false;

  @State()
  private hasError: boolean = false;

  @State()
  private hasHelp: boolean = false;

  @State()
  private hasLabelInfo: boolean = false;

  @Watch('hasError')
  watchValue(hasError: boolean): void {
    const phoneInputSlot = this.root.querySelector('gux-phone-input-beta');
    if (phoneInputSlot) {
      phoneInputSlot.hasError = hasError;
    }
  }

  @OnMutation({ childList: true, subtree: true })
  onMutation(): void {
    this.labelInfo = this.root.querySelector('[slot=label-info]');
    this.hasError = hasSlot(this.root, 'error');
    this.hasHelp = hasSlot(this.root, 'help');
    this.hasLabelInfo = hasSlot(this.root, 'label-info');
  }

  @Listen('keyup')
  handleKeyup(event: KeyboardEvent): void {
    switch (event.key) {
      case 'Tab': {
        if (this.input.matches(':focus-within')) {
          void this.labelInfo?.showTooltip();
          this.hideLabelInfoTimeout = setTimeout(() => {
            void this.labelInfo?.hideTooltip();
          }, 6000);
        }
        break;
      }
      default: {
        if (this.input.matches(':focus-within')) {
          void this.labelInfo?.hideTooltip();
          clearTimeout(this.hideLabelInfoTimeout);
        }
        break;
      }
    }
  }

  @Listen('focusout')
  onFocusout(): void {
    void this.labelInfo?.hideTooltip();
    clearTimeout(this.hideLabelInfoTimeout);
  }

  async componentWillLoad(): Promise<void> {
    this.getI18nValue = await buildI18nForComponent(
      this.root,
      componentResources as ILocalizedComponentResources
    );

    this.setInput();
    this.setLabel();

    this.labelInfo = this.root.querySelector('[slot=label-info]');
    this.hasError = hasSlot(this.root, 'error');
    this.hasHelp = hasSlot(this.root, 'help');
    this.hasLabelInfo = hasSlot(this.root, 'label-info');

    trackComponent(this.root, { variant: this.variant });
  }

  disconnectedCallback(): void {
    this.disabledObserver.disconnect();
    this.requiredObserver.disconnect();
  }

  render(): JSX.Element {
    return (
      <GuxFormFieldFieldsetContainer labelPosition={this.computedLabelPosition}>
        <GuxFormFieldScreenreaderLabel>
          {this.label?.textContent}
          {this.renderText(this.getI18nValue('required'), this.required)}
          {this.renderText(
            getSlotTextContent(this.root, 'error'),
            this.hasError
          )}
          {this.renderText(
            getSlotTextContent(this.root, 'label-info'),
            this.hasLabelInfo
          )}
        </GuxFormFieldScreenreaderLabel>
        <GuxFormFieldVisualLabel
          position={this.computedLabelPosition}
          required={this.required}
        >
          <slot name="label" onSlotchange={() => this.setLabel()} />
          <gux-form-field-label-indicator
            variant={this.indicatorMark}
            required={this.required}
          />
          <slot name="label-info"></slot>
        </GuxFormFieldVisualLabel>
        <div class="gux-input-and-error-container">
          <div
            class={{
              'gux-input': true,
              'gux-input-error': this.hasError
            }}
          >
            <div
              class={{
                'gux-input-container': true,
                'gux-disabled': this.disabled
              }}
            >
              <slot />
            </div>
          </div>
          <GuxFormFieldError show={this.hasError}>
            <slot name="error" />
          </GuxFormFieldError>
          <GuxFormFieldHelp show={!this.hasError && this.hasHelp}>
            <slot name="help" />
          </GuxFormFieldHelp>
        </div>
      </GuxFormFieldFieldsetContainer>
    ) as JSX.Element;
  }

  private renderText(text: string, condition: boolean = false): string {
    if (condition) {
      return ' ' + text;
    }
  }

  private get variant(): string {
    const labelPositionVariant = this.labelPosition
      ? this.labelPosition.toLowerCase()
      : 'none';

    const type = 'phoneInput';

    return `${type}-${labelPositionVariant}`;
  }

  private setInput(): void {
    this.input = this.root.querySelector('gux-phone-input-beta');

    this.disabled = this.input.disabled;
    this.required = this.input.required;

    this.disabledObserver = onDisabledChange(
      this.input,
      (disabled: boolean) => {
        this.disabled = disabled;
      }
    );
    this.requiredObserver = onRequiredChange(
      this.input,
      (required: boolean) => {
        this.required = required;
      }
    );

    validateFormIds(this.root, this.input);
  }

  private setLabel(): void {
    this.label = this.root.querySelector('label[slot="label"]');

    this.computedLabelPosition = getComputedLabelPosition(
      this.label,
      this.labelPosition
    );
  }
}
