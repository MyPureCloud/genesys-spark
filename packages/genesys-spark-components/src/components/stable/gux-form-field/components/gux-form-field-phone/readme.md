# gux-form-field-phone



<!-- Auto Generated Below -->


## Properties

| Property        | Attribute        | Description                                                                                                                                                                                                                                                                     | Type                                    | Default      |
| --------------- | ---------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | --------------------------------------- | ------------ |
| `indicatorMark` | `indicator-mark` | Field indicator mark which can show *, (optional) or blank Defaults to required. When set to required, the component will display * for required fields and blank for optional When set to optional, the component will display (optional) for optional and blank for required. | `"optional" \| "required"`              | `'required'` |
| `labelPosition` | `label-position` |                                                                                                                                                                                                                                                                                 | `"above" \| "beside" \| "screenreader"` | `undefined`  |


## Slots

| Slot           | Description                                |
| -------------- | ------------------------------------------ |
|                | Required slot for gux-phone-input-beta tag |
| `"error"`      | Optional slot for error message            |
| `"help"`       | Optional slot for help message             |
| `"label"`      | Required slot for label tag                |
| `"label-info"` | Optional slot for label tooltip            |


## Dependencies

### Depends on

- [gux-form-field-label-indicator](../../helper-components/gux-form-field-optional-indicator)
- [gux-icon](../../../gux-icon)

### Graph
```mermaid
graph TD;
  gux-form-field-phone --> gux-form-field-label-indicator
  gux-form-field-phone --> gux-icon
  style gux-form-field-phone fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
