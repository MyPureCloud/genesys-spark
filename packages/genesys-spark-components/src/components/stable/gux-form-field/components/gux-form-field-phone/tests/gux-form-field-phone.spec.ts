import { newSpecPage } from '@test/specTestUtils';

import { GuxFormFieldPhone } from '../gux-form-field-phone';

const components = [GuxFormFieldPhone];
const language = 'en';

describe('gux-form-field-phone', () => {
  it('should build', async () => {
    const html = `
      <gux-form-field-phone>
        <gux-phone-input-beta></gux-phone-input-beta>
        <label slot="label">Label</label>
      </gux-form-field-phone>
    `;
    const page = await newSpecPage({ components, html, language });

    expect(page.rootInstance).toBeInstanceOf(GuxFormFieldPhone);
  });

  describe('#render', () => {
    describe('label-position', () => {
      [
        '',
        'label-position="above"',
        'label-position="beside"',
        'label-position="screenreader"'
      ].forEach((componentAttribute, index) => {
        it(`should render component as expected (${index + 1})`, async () => {
          const html = `
            <gux-form-field-phone ${componentAttribute}>
              <gux-phone-input-beta></gux-phone-input-beta>
              <label slot="label">Label</label>
            </gux-form-field-phone>
          `;
          const page = await newSpecPage({ components, html, language });

          expect(page.root).toMatchSnapshot();
        });
      });
    });

    describe('error', () => {
      it('should render component as expected', async () => {
        const html = `
        <gux-form-field-phone>
          <gux-phone-input-beta></gux-phone-input-beta>
          <label slot="label">Default</label>
          <span slot="error">Custom Error Message</span>
        </gux-form-field-phone>
        `;
        const page = await newSpecPage({ components, html, language });

        expect(page.root).toMatchSnapshot();
      });
    });

    describe('help', () => {
      it('should render component as expected', async () => {
        const html = `
        <gux-form-field-phone>
          <gux-phone-input-beta></gux-phone-input-beta>
          <label slot="label">Default</label>
          <span slot="help">Custom Help Message</span>
        </gux-form-field-phone>
        `;
        const page = await newSpecPage({ components, html, language });

        expect(page.root).toMatchSnapshot();
      });
    });
  });
});
