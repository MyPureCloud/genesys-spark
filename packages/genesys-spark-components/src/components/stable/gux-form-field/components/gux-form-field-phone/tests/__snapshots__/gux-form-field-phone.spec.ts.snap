// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gux-form-field-phone #render error should render component as expected 1`] = `
<gux-form-field-phone>
  <mock:shadow-root>
    <fieldset class="gux-above gux-form-field-fieldset-container">
      <legend class="gux-form-field-legend-label gux-screenreader">
        <slot name="label"></slot>
        <gux-screen-reader-beta>
          Required
        </gux-screen-reader-beta>
        <gux-screen-reader-beta>
          Custom Error Message
        </gux-screen-reader-beta>
      </legend>
      <span aria-hidden="true" class="gux-above gux-form-field-legend-label" role="presentation">
        Default
      </span>
      <div class="gux-input-and-error-container">
        <div class="gux-input gux-input-error">
          <div class="gux-input-container">
            <slot></slot>
          </div>
        </div>
        <div class="gux-form-field-error gux-show">
          <gux-icon decorative="" icon-name="fa/hexagon-exclamation-solid"></gux-icon>
          <div class="gux-message">
            <slot name="error"></slot>
          </div>
        </div>
        <div class="gux-form-field-help">
          <div class="gux-message">
            <slot name="help"></slot>
          </div>
        </div>
      </div>
    </fieldset>
  </mock:shadow-root>
  <gux-phone-input-beta aria-describedby="gux-form-field-error-i" id="gux-form-field-input-i"></gux-phone-input-beta>
  <label for="gux-form-field-input-i" slot="label">
    Default
  </label>
  <span id="gux-form-field-error-i" slot="error">
    Custom Error Message
  </span>
</gux-form-field-phone>
`;

exports[`gux-form-field-phone #render help should render component as expected 1`] = `
<gux-form-field-phone>
  <mock:shadow-root>
    <fieldset class="gux-above gux-form-field-fieldset-container">
      <legend class="gux-form-field-legend-label gux-screenreader">
        <slot name="label"></slot>
        <gux-screen-reader-beta>
          Required
        </gux-screen-reader-beta>
      </legend>
      <span aria-hidden="true" class="gux-above gux-form-field-legend-label" role="presentation">
        Default
      </span>
      <div class="gux-input-and-error-container">
        <div class="gux-input">
          <div class="gux-input-container">
            <slot></slot>
          </div>
        </div>
        <div class="gux-form-field-error">
          <gux-icon decorative="" icon-name="fa/hexagon-exclamation-solid"></gux-icon>
          <div class="gux-message">
            <slot name="error"></slot>
          </div>
        </div>
        <div class="gux-form-field-help gux-show">
          <div class="gux-message">
            <slot name="help"></slot>
          </div>
        </div>
      </div>
    </fieldset>
  </mock:shadow-root>
  <gux-phone-input-beta aria-describedby="gux-form-field-help-i" id="gux-form-field-input-i"></gux-phone-input-beta>
  <label for="gux-form-field-input-i" slot="label">
    Default
  </label>
  <span id="gux-form-field-help-i" slot="help">
    Custom Help Message
  </span>
</gux-form-field-phone>
`;

exports[`gux-form-field-phone #render label-position should render component as expected (1) 1`] = `
<gux-form-field-phone>
  <mock:shadow-root>
    <fieldset class="gux-above gux-form-field-fieldset-container">
      <legend class="gux-form-field-legend-label gux-screenreader">
        <slot name="label"></slot>
        <gux-screen-reader-beta>
          Required
        </gux-screen-reader-beta>
      </legend>
      <span aria-hidden="true" class="gux-above gux-form-field-legend-label" role="presentation">
        Label
      </span>
      <div class="gux-input-and-error-container">
        <div class="gux-input">
          <div class="gux-input-container">
            <slot></slot>
          </div>
        </div>
        <div class="gux-form-field-error">
          <gux-icon decorative="" icon-name="fa/hexagon-exclamation-solid"></gux-icon>
          <div class="gux-message">
            <slot name="error"></slot>
          </div>
        </div>
        <div class="gux-form-field-help">
          <div class="gux-message">
            <slot name="help"></slot>
          </div>
        </div>
      </div>
    </fieldset>
  </mock:shadow-root>
  <gux-phone-input-beta id="gux-form-field-input-i"></gux-phone-input-beta>
  <label for="gux-form-field-input-i" slot="label">
    Label
  </label>
</gux-form-field-phone>
`;

exports[`gux-form-field-phone #render label-position should render component as expected (2) 1`] = `
<gux-form-field-phone label-position="above">
  <mock:shadow-root>
    <fieldset class="gux-above gux-form-field-fieldset-container">
      <legend class="gux-form-field-legend-label gux-screenreader">
        <slot name="label"></slot>
        <gux-screen-reader-beta>
          Required
        </gux-screen-reader-beta>
      </legend>
      <span aria-hidden="true" class="gux-above gux-form-field-legend-label" role="presentation">
        Label
      </span>
      <div class="gux-input-and-error-container">
        <div class="gux-input">
          <div class="gux-input-container">
            <slot></slot>
          </div>
        </div>
        <div class="gux-form-field-error">
          <gux-icon decorative="" icon-name="fa/hexagon-exclamation-solid"></gux-icon>
          <div class="gux-message">
            <slot name="error"></slot>
          </div>
        </div>
        <div class="gux-form-field-help">
          <div class="gux-message">
            <slot name="help"></slot>
          </div>
        </div>
      </div>
    </fieldset>
  </mock:shadow-root>
  <gux-phone-input-beta id="gux-form-field-input-i"></gux-phone-input-beta>
  <label for="gux-form-field-input-i" slot="label">
    Label
  </label>
</gux-form-field-phone>
`;

exports[`gux-form-field-phone #render label-position should render component as expected (3) 1`] = `
<gux-form-field-phone label-position="beside">
  <mock:shadow-root>
    <fieldset class="gux-beside gux-form-field-fieldset-container">
      <legend class="gux-form-field-legend-label gux-screenreader">
        <slot name="label"></slot>
        <gux-screen-reader-beta>
          Required
        </gux-screen-reader-beta>
      </legend>
      <span aria-hidden="true" class="gux-beside gux-form-field-legend-label" role="presentation">
        Label
      </span>
      <div class="gux-input-and-error-container">
        <div class="gux-input">
          <div class="gux-input-container">
            <slot></slot>
          </div>
        </div>
        <div class="gux-form-field-error">
          <gux-icon decorative="" icon-name="fa/hexagon-exclamation-solid"></gux-icon>
          <div class="gux-message">
            <slot name="error"></slot>
          </div>
        </div>
        <div class="gux-form-field-help">
          <div class="gux-message">
            <slot name="help"></slot>
          </div>
        </div>
      </div>
    </fieldset>
  </mock:shadow-root>
  <gux-phone-input-beta id="gux-form-field-input-i"></gux-phone-input-beta>
  <label for="gux-form-field-input-i" slot="label">
    Label
  </label>
</gux-form-field-phone>
`;

exports[`gux-form-field-phone #render label-position should render component as expected (4) 1`] = `
<gux-form-field-phone label-position="screenreader">
  <mock:shadow-root>
    <fieldset class="gux-form-field-fieldset-container gux-screenreader">
      <legend class="gux-form-field-legend-label gux-screenreader">
        <slot name="label"></slot>
        <gux-screen-reader-beta>
          Required
        </gux-screen-reader-beta>
      </legend>
      <span aria-hidden="true" class="gux-form-field-legend-label gux-screenreader" role="presentation">
        Label
      </span>
      <div class="gux-input-and-error-container">
        <div class="gux-input">
          <div class="gux-input-container">
            <slot></slot>
          </div>
        </div>
        <div class="gux-form-field-error">
          <gux-icon decorative="" icon-name="fa/hexagon-exclamation-solid"></gux-icon>
          <div class="gux-message">
            <slot name="error"></slot>
          </div>
        </div>
        <div class="gux-form-field-help">
          <div class="gux-message">
            <slot name="help"></slot>
          </div>
        </div>
      </div>
    </fieldset>
  </mock:shadow-root>
  <gux-phone-input-beta id="gux-form-field-input-i"></gux-phone-input-beta>
  <label for="gux-form-field-input-i" slot="label">
    Label
  </label>
</gux-form-field-phone>
`;
